[Unified Text Replacement] Replacements should maintain the original text attributes

rr-codes opened the pull request:
#### 6cffe2c43cbad3173e4431d13a8c37881ad218fc
<pre>
[Unified Text Replacement] Replacements should maintain the original text attributes
<a href="https://bugs.webkit.org/show_bug.cgi?id=270606">https://bugs.webkit.org/show_bug.cgi?id=270606</a>
<a href="https://rdar.apple.com/122835651">rdar://122835651</a>

Reviewed by Wenson Hsieh.

Add support for attributed strings to work properly and have their attributes persisted.

Also, add `WEBCORE_EXPORT` to `showTreeForThis` to ease future debugging, and make a correctness
fix for an unrelated API test.

* Source/WebCore/dom/Node.h:
* Source/WebCore/editing/WebContentReader.h:
* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::Editor::replaceSelectionWithAttributedString):
* Source/WebCore/editing/cocoa/WebContentReaderCocoa.mm:
(WebCore::createFragment):
(WebCore::WebContentReader::readRTFD):
(WebCore::WebContentMarkupReader::readRTFD):
(WebCore::WebContentReader::readRTF):
(WebCore::WebContentMarkupReader::readRTF):
(WebCore::createFragmentAndAddResources): Deleted.
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm: Renamed from Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp.
(WebKit::replaceTextInRange):
(WebKit::replaceContentsInRange):
(WebKit::extendedBoundaryPoint):
(WebKit::findReplacementMarkerByUUID):
(WebKit::UnifiedTextReplacementController::UnifiedTextReplacementController):
(WebKit::UnifiedTextReplacementController::willBeginTextReplacementSession):
(WebKit::UnifiedTextReplacementController::didBeginTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveReplacements):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidUpdateStateForReplacement):
(WebKit::UnifiedTextReplacementController::didEndTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveTextWithReplacementRange):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveEditAction):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CopyRTF.mm:
(checkColor):

Canonical link: <a href="https://commits.webkit.org/275821@main">https://commits.webkit.org/275821@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ba0c4ae603db429bafb1ee71671fd9670f7ce388

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42958 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21982 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45360 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45580 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39083 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45264 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25694 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19406 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43531 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19005 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36986 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16518 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16607 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/38050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1014 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39132 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38378 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47106 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17805 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14659 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/42298 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/19410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/40954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9571 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19588 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/19040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/dom/Node.h:
```diff
@@ -488,7 +488,7 @@ class Node : public EventTarget {
 #if ENABLE(TREE_DEBUGGING)
     void showNode(const char* prefix = "") const;
-    void showTreeForThis() const;
+    WEBCORE_EXPORT void showTreeForThis() const;
```
This change should be undone.

rr-codes replied:
why exactly would it be bad to leave it in? I find myself often adding it for debugging, which results in a mostly full-rebuil because of the header it's in

pxlcoder replied:
I think that's a good reason to keep this! Can you mention that in the ChangeLog?

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/CopyRTF.mm:
```diff
@@ -89,7 +89,7 @@ static void checkColor(PlatformColor *color, CGFloat red, CGFloat green, CGFloat
 
     EXPECT_EQ(red, observedRed);
-    EXPECT_EQ(green, observedRed);
+    EXPECT_EQ(green, observedGreen);
```
Was this just working before because the tested red and green values were always equal?

rr-codes replied:
yup lol

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -338,41 +354,37 @@ void UnifiedTextReplacementController::textReplacementSessionDidReceiveTextWithR
+    }
 
-    auto newResolvedRange = resolveCharacterRange(*sessionRange, newRange);
+    RELEASE_ASSERT(sessionRangeCharacterCount >= range.location + range.length);
```
is this release assert ok to leave in? It's needed because otherwise there would be an underflow in the below lines if this was somehow ever false. I could do an early return with a debug assert too, but this should never ever happen in the first place.

pxlcoder replied:
A `RELEASE_ASSERT` is not the answer here, unless there is some security implication. Otherwise, why are we creating an even worse user experience because of a bug in our code?

whsieh commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -338,41 +354,41 @@ void UnifiedTextReplacementController::textReplacementSessionDidReceiveTextWithR
 
-    replaceTextInRange(*frame, newResolvedRange, newText);
+    if (UNLIKELY(sessionRangeCharacterCount < range.location + range.length)) {
+        RELEASE_LOG(UnifiedTextReplacement, "UnifiedTextReplacementController::textReplacementSessionDidReceiveTextWithReplacementRange (%s) => trying to replace a range larger than the context range (context range count: %llu, range.location %llu, range.length %llu)", uuid.toString().utf8().data(), sessionRangeCharacterCount, range.location, range.length);
```
Nit - `RELEASE_LOG_ERROR`?

whsieh commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -470,6 +484,24 @@ void UnifiedTextReplacementController::textReplacementSessionDidReceiveEditActio
+    case WebKit::WebTextReplacementData::EditAction::UndoAll: {
+        auto updatedContents = updatedLiveRange->cloneContents();
+        if (updatedContents.hasException()) {
+            RELEASE_LOG(UnifiedTextReplacement, "UnifiedTextReplacementController::textReplacementSessionDidReceiveEditAction (%s) => exception when cloning contents after action", uuid.toString().utf8().data());
```
Ditto

rr-codes committed:
[Unified Text Replacement] Replacements should maintain the original text attributes
https://bugs.webkit.org/show_bug.cgi?id=270606
rdar://122835651

Reviewed by Wenson Hsieh.

Add support for attributed strings to work properly and have their attributes persisted.

Also, add `WEBCORE_EXPORT` to `showTreeForThis` to ease future debugging, and make a correctness
fix for an unrelated API test.

* Source/WebCore/dom/Node.h:
* Source/WebCore/editing/WebContentReader.h:
* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::Editor::replaceSelectionWithAttributedString):
* Source/WebCore/editing/cocoa/WebContentReaderCocoa.mm:
(WebCore::createFragment):
(WebCore::WebContentReader::readRTFD):
(WebCore::WebContentMarkupReader::readRTFD):
(WebCore::WebContentReader::readRTF):
(WebCore::WebContentMarkupReader::readRTF):
(WebCore::createFragmentAndAddResources): Deleted.
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm: Renamed from Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp.
(WebKit::replaceTextInRange):
(WebKit::replaceContentsInRange):
(WebKit::extendedBoundaryPoint):
(WebKit::findReplacementMarkerByUUID):
(WebKit::UnifiedTextReplacementController::UnifiedTextReplacementController):
(WebKit::UnifiedTextReplacementController::willBeginTextReplacementSession):
(WebKit::UnifiedTextReplacementController::didBeginTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveReplacements):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidUpdateStateForReplacement):
(WebKit::UnifiedTextReplacementController::didEndTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveTextWithReplacementRange):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveEditAction):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CopyRTF.mm:
(checkColor):

Canonical link: https://commits.webkit.org/275821@main
